### PR TITLE
feat(button)!: migrating button to core-tokens (CSS-97)

### PR DESCRIPTION
**Pull request by bernhard-adobe**
Migrating the Button to core-tokens.

## Description

- removed skin.css
- added all the theming for Spectrum and Express in the `themes` folder
- upped version to 4.0.0 as core-tokens are a major change, updated CHANGELOG.md
- added temporary font-weights into `components/tokens/custom-spectrum/custom-vars.css`
- followed action-button example for migration and this doc https://github.com/adobe/spectrum-css/wiki/Migrating-a-component-to-core-tokens

Pages:
https://opensource.adobe.com/spectrum-css/button-accent.html

https://opensource.adobe.com/spectrum-css/button-negative.html

https://opensource.adobe.com/spectrum-css/button-primary.html

https://opensource.adobe.com/spectrum-css/button-secondary.html

https://opensource.adobe.com/spectrum-css/button-staticcolor.html

Source-file for migration: Components batch 1.xd

https://jira.corp.adobe.com/browse/CSS-97

*Spectrum Web Components:*
https://github.com/adobe/spectrum-web-components/pull/2659 feat(button): using core-tokens for button #2659

## How and where has this been tested?
 - **How this was tested:** Google Chrome natively running `npm run dev`, compared to the opensource website
 - **Browser(s) and OS(s) this was tested with:** Version 106.0.5200.0 (Official Build) canary (arm64)

## Screenshots
Note: the color for the button label changed from transparent to a solid color. This is documented in the XD source file.
Screen Shot 2022-07-25 at 12 36 32

Screen Shot 2022-07-22 at 14 15 03

## To-do list

- [ ] If my change impacts other components, I have tested to make sure they don't break.
- [x] If my change impacts documentation, I have updated the documentation accordingly.
- [x] I have read the [CONTRIBUTING document](/.github/CONTRIBUTING.md).
- [ ] I have tested these changes in Windows High Contrast mode.

- [ ] This pull request is ready to merge.

### CODE REVIEW

**Comment by bernhard-adobe** [comment]
VRT running manually: https://spectrum-visual-regression.ci.corp.adobe.com/view/Spectrum%20CSS/job/css-vrt-test/10/

**Comment by bernhard-adobe** [comment]
Looks like some more adjustments needed for min-width:
https://spectrum-visual-regression.ci.corp.adobe.com/view/Spectrum%20CSS/job/css-vrt-test/12/artifact/backstop_data/html_report/index.html
Screen Shot 2022-07-25 at 15 45 49

File: components/button/index.css [hunk 1]
```diff
@@ -13,76 +13,109 @@ governing permissions and limitations under the License.
 @import "../vars/css/scales/spectrum-large.css";
 @import "../vars/css/scales/spectrum-medium.css";
 @import "../commons/basebutton.css";
```
Comment (pfulton) [inline components/button/index.css:15]: We'll need to remove these lines, as we're no longer using the `vars` package.

File: components/commons/package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@spectrum-css/commons",
-  "version": "3.0.6",
+  "version": "4.0.0",
```
Comment (pfulton) [inline components/commons/package.json:3]: No need to increment this manually. Lerna will take care of this for us automatically when we do the release.

File: components/commons/CHANGELOG.md [hunk 3]
```diff
@@ -3,6 +3,14 @@
...
 See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.
 
+
+<a name="3.0.6"></a>
+## 4.0.0
+🗓 2022-06-16 • 📝 [Commits](https://github.com/adobe/spectrum-css/compare/@spectrum-css/commons@3.0.5...@spectrum-css/commons@4.0.0)
+
+**Note:** Version bump for core tokens
+
+
```
Comment (pfulton) [inline components/commons/CHANGELOG.md:13]: No need to update this manually. Lerna will take care of this automatically when we do the release, and the messages that it puts in here are based on your conventional commit(s)

File: components/button/package.json [hunk 4]
```diff
@@ -19,7 +19,7 @@
     "@spectrum-css/vars": "^8.0.0"
```
Comment (pfulton) [inline components/button/package.json:19]: Let's remove this line, as we no longer need the `vars` package.

File: components/button/package.json [hunk 5]
```diff
@@ -19,7 +19,7 @@
...
   },
   "devDependencies": {
-    "@spectrum-css/component-builder": "^3.2.0",
+    "@spectrum-css/component-builder-simple": "^1.0.0-beta.0",
     "@spectrum-css/icon": "^3.0.23",
     "@spectrum-css/vars": "^8.0.0",
```
Comment (pfulton) [inline components/button/package.json:24]: We'll remove this line, too. Instead, we need to require the tokens package.

`"@spectrum-css/tokens": "^1.0.2"`

Please also make the tokens package a `peerDependency`.

File: components/button/index.css [hunk 6]
```diff
@@ -110,11 +145,11 @@ governing permissions and limitations under the License.
     padding-inline-end: 0;
   }
 
-  --spectrum-button-focus-ring-color: var(--spectrum-button-m-primary-fill-texticon-focus-ring-color-key-focus);
+  /* --spectrum-button-focus-ring-color: var(--spectrum-button-m-primary-fill-texticon-focus-ring-color-key-focus); */
```
Comment (pfulton) [inline components/button/index.css:148]: Are we able to remove this commented out line now?

**Review by pfulton — CHANGES_REQUESTED** [review]
Nice work, @bernhard-adobe! Thanks for taking care of this!

I left a few comments here that I think we should address. Additionally, can you go through and be sure that you've added the `--mod-` prefixed custom properties in any of the places where you're using a `--spectrum-` custom property? I noticed there were a few that were missed.

File: components/button/index.css (hunk 1 above)
Comment (bernhard-adobe) [reply components/button/index.css:15]: removed

File: components/button/index.css (hunk 6 above)
Comment (bernhard-adobe) [reply components/button/index.css:146]: yes, we get a good focus ring color on all button variants.

File: components/button/index.css [hunk 7]
```diff
@@ -13,76 +13,109 @@ governing permissions and limitations under the License.
...
+  position: relative;
+
 
   padding-block-start: 0;
   padding-block-end: 0;
 
   /* Start with text-only padding */
-  padding-inline: var(--spectrum-button-primary-fill-textonly-padding-left-adjusted) var(--spectrum-button-primary-fill-textonly-padding-right-adjusted);
+  /* padding-inline-start: var(--mod-button-edge-to-text, var(--spectrum-button-edge-to-text));
+  padding-inline-end: var(--mod-button-edge-to-text, var(--spectrum-button-edge-to-text)); */
```
[the remark this answers is not in the document]
Comment (bernhard-adobe) [reply components/button/index.css:112]: maybe, lets see if the VRT picks up the padding now correctly. there were some issues in https://spectrum-visual-regression.ci.corp.adobe.com/view/Spectrum%20CSS/job/css-vrt-test/12/artifact/backstop_data/html_report/index.html

**Comment by bernhard-adobe** [comment]
new VRT is: https://spectrum-visual-regression.ci.corp.adobe.com/view/Spectrum%20CSS/job/css-vrt-test/17/

**Comment by bernhard-adobe** [comment]
never-mind, we got https://spectrum-visual-regression.ci.corp.adobe.com/view/Spectrum%20CSS/job/css-vrt-test/21/ now to include the changes from Jians cleanup-script  #1487

**Comment by bernhard-adobe** [comment]
There are visual differences with the min-width of the buttons. Need to investigate what went wrong or missing from the migration.
Screen Shot 2022-07-29 at 14 43 12

**Comment by bernhard-adobe** [comment]
PJ provided us with the correct 2.25 ratio for button-min-width against height.

https://spectrum-visual-regression.ci.corp.adobe.com/view/Spectrum%20CSS/job/css-vrt-test/22/

Rule: https://spectrum.corp.adobe.com/page/button/#Minimum-width

**Comment by bernhard-adobe** [comment]
@pfulton VRT now looks as expected. We get some more pixels for button in min-width:
Screen Shot 2022-08-01 at 17 23 27

File: components/button/index.css [hunk 8]
```diff
@@ -10,79 +10,117 @@ OF ANY KIND, either express or implied. See the License for the specific languag
 governing permissions and limitations under the License.
 */
 
-@import "../vars/css/scales/spectrum-large.css";
-@import "../vars/css/scales/spectrum-medium.css";
 @import "../commons/basebutton.css";
```
Comment (pfulton) [inline components/button/index.css:13]: I think we need to remove this line and only import the `basebutton-coretokens.css` file.

File: components/button/index.css [hunk 9]
```diff
@@ -93,15 +131,17 @@ governing permissions and limitations under the License.
   color: inherit;
 
   .spectrum-Icon {
-    @inherit: %spectrum-ButtonIcon;
+    inset-inline-end: calc(var(--mod-spectrum-button-edge-to-visual, var(--spectrum-button-edge-to-visual)) - var(--mod-button-border-width, var(--spectrum-button-border-width)));
+    inset-block-end: calc(var(--mod-spectrum-button-edge-to-visual, var(--spectrum-button-edge-to-visual)) - var(--mod-button-border-width, var(--spectrum-button-border-width)));
```
Comment (pfulton) [inline components/button/index.css:135]: I think we should alias this calculation to its own custom property: it's the same for both properties (yes?) and it'll make things a bit cleaner.

**Review by pfulton — CHANGES_REQUESTED** [review]
Thanks for tackling this one, @bernhard-adobe. I know this is a pretty big component!

It looks like there might be something funky going on with this branch's history. I took a look at the relevant `Button` files, and left a few comments here.

Also, can we bring over the Windows High Contrast styles that were in the `skin.css` file previously? Thanks!

**Comment by bernhard-adobe** [comment]
@pfulton added the correct high-contrast mode updates for the buttons, including the accent variant.
Thanks for the hint about `forced-color-adjust: none;`

Screen Shot 2022-08-17 at 23 31 28


Screen Shot 2022-08-17 at 16 25 07
321/185302683-b6b36159-20a6-4a4a-adce-952a6e3c5f13.png">

**Comment by bernhard-adobe** [comment]
@pfulton  I could quickly resolve the merge conflicts tomorrow if the PR look good to you!

**Comment by bernhard-adobe** [comment]
> @bernhard-adobe yeah, can you resolve the conflicts please? After that, I'll do the beta release. Thanks!

@pfulton Easy enough. Done. Thank you

**Comment by pfulton** [comment]
Beta released:
`7.0.0-beta.0`

**Comment by pfulton** [comment]
@bernhard-adobe Can you open a PR in the [Spectrum Web Components repo](https://github.com/adobe/spectrum-web-components/pulls) with this update?

**Comment by pfulton** [comment]
Released a new beta: `@spectrum-css/button@7.0.0-beta.3`

**Comment by castastrophe** [comment]
New beta release: @spectrum-css/button@7.0.0-beta.4

**Comment by castastrophe** [comment]
Apologies for the delay! I've published the latest beta releases:

 - @spectrum-css/button@7.0.0-beta.5
 - @spectrum-css/commons@4.1.0-beta.0
 

<details>
<summary>Full output</summary>

```bash
lerna success published @spectrum-css/button 7.0.0-beta.5
lerna notice 
lerna notice 📦  @spectrum-css/button@7.0.0-beta.5
lerna notice === Tarball Contents === 
lerna notice 11.4kB LICENSE                        
lerna notice 29.8kB dist/themes/express.css        
lerna notice 1.8kB  themes/express.css             
lerna notice 50.0kB dist/index-base.css            
lerna notice 31.0kB dist/index-theme.css           
lerna notice 81.0kB dist/index-vars.css            
lerna notice 81.0kB dist/index.css                 
lerna notice 13.0kB index.css                      
lerna notice 29.8kB dist/themes/spectrum.css       
lerna notice 28.3kB themes/spectrum.css            
lerna notice 4.0kB  stories/button.stories.js      
lerna notice 68B    gulpfile.js                    
lerna notice 1.2kB  stories/template.js            
lerna notice 842B   package.json                   
lerna notice 21.6kB CHANGELOG.md                   
lerna notice 314B   README.md                      
lerna notice 7.5kB  metadata/button-accent.yml     
lerna notice 9.1kB  metadata/button-negative.yml   
lerna notice 8.9kB  metadata/button-primary.yml    
lerna notice 9.0kB  metadata/button-secondary.yml  
lerna notice 21.7kB metadata/button-staticcolor.yml
lerna notice === Tarball Details === 
lerna notice name:          @spectrum-css/button                    
lerna notice version:       7.0.0-beta.5                            
lerna notice filename:      spectrum-css-button-7.0.0-beta.5.tgz    
lerna notice package size:  36.5 kB                                 
lerna notice unpacked size: 441.1 kB                                
lerna notice shasum:        75c5ba0fad52cfc776c7e79b9e271e90b66d1b28
lerna notice integrity:     sha512-wTyhJ0zt69eIE[...]rg3w/4yeXfeBw==
lerna notice total files:   21                                      
lerna notice 
lerna http fetch PUT 200 https://registry.npmjs.org/@spectrum-css%2fcommons 1328ms
lerna success published @spectrum-css/commons 4.1.0-beta.0
lerna notice 
lerna notice 📦  @spectrum-css/commons@4.1.0-beta.0
lerna notice === Tarball Contents === 
lerna notice 11.4kB LICENSE                  
lerna notice 3.5kB  basebutton-coretokens.css
lerna notice 3.4kB  basebutton.css           
lerna notice 1.8kB  overlay-coretokens.css   
lerna notice 626B   package.json             
lerna notice 6.2kB  CHANGELOG.md             
lerna notice 337B   README.md                
lerna notice === Tarball Details === 
lerna notice name:          @spectrum-css/commons                   
lerna notice version:       4.1.0-beta.0                            
lerna notice filename:      spectrum-css-commons-4.1.0-beta.0.tgz   
lerna notice package size:  7.3 kB                                  
lerna notice unpacked size: 27.3 kB                                 
lerna notice shasum:        d7d3c6a0e583e082a2923e6e692bdce07b85ac73
lerna notice integrity:     sha512-9t+5EIlukpTG1[...]kSzmVcBTlvalg==
lerna notice total files:   7                                       
lerna notice 
Successfully published:
 - @spectrum-css/button@7.0.0-beta.5
 - @spectrum-css/commons@4.1.0-beta.0
lerna success published 2 packages
```
</details>

**Comment by bernhard-adobe** [comment]
> Apologies for the delay! I've published the latest beta releases:
> 
> * @spectrum-css/button@7.0.0-beta.5
> * @spectrum-css/commons@4.1.0-beta.0
> 
> Full output
All good. @castastrophe 
That unfortunately didn't fix the visual differences in Web components either.

**Review by pfulton — APPROVED** [review]

**Pull request merged**